File: python/verlet.py

import numpy as np
from numba import njit
from forces import compute_accelerations, lennard_jones_potential
from typing import Tuple


def velocity_verlet_step(positions: np.ndarray,
                        velocities: np.ndarray,
                        accelerations: np.ndarray,
                        dt: float,
                        cell_size: float,
                        r_cut: float = 2.5,
                        **kwargs) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:

    a_old = accelerations.copy()
    new_positions = positions + velocities * dt + 0.5 * a_old * dt**2
    new_accelerations = compute_accelerations(new_positions, cell_size, r_cut)
    new_velocities = velocities + 0.5 * (a_old + new_accelerations) * dt

    mask1 = new_positions > cell_size
    mask2 = new_positions < 0
    new_positions[mask1] -= cell_size
    new_positions[mask2] += cell_size

    return new_positions, new_velocities, new_accelerations


def position_verlet_step(positions: np.ndarray,
                        positions_prev: np.ndarray,
                        accelerations: np.ndarray,
                        dt: float,
                        cell_size: float,
                        **kwargs) -> Tuple[np.ndarray, np.ndarray]:

    disp = positions - positions_prev
    disp -= cell_size * np.round(disp / cell_size)

    new_positions = positions + disp + accelerations * dt**2

    new_disp = new_positions - positions
    velocities = (disp + new_disp) / (2 * dt)

    new_positions = new_positions % cell_size

    return new_positions, velocities


def kinetic_energy(velocities: np.ndarray) -> float:
    return 0.5 * np.sum(velocities ** 2)


@njit
def potential_energy(positions: np.ndarray, cell_size: float, r_cut: float = 2.5) -> float:
    N = len(positions)
    rc = min(r_cut, cell_size / 2)
    u_shift = lennard_jones_potential(rc)
    pe = 0.0
    for i in range(N):
        for j in range(i + 1, N):
            r_vec = positions[j] - positions[i]
            for k in range(3):
                if r_vec[k] > cell_size / 2:
                    r_vec[k] -= cell_size
                if r_vec[k] < -cell_size / 2:
                    r_vec[k] += cell_size
            r = np.sqrt(r_vec[0]**2 + r_vec[1]**2 + r_vec[2]**2)
            if r < rc:
                pe += lennard_jones_potential(r) - u_shift
    return pe

File: src/verlet.cpp
#include "verlet.h"
#include <cmath>
#include <algorithm>

void velocity_verlet_step(std::vector<Vec3>& pos,
                          std::vector<Vec3>& vel,
                          std::vector<Vec3>& acc,
                          double dt,
                          double cell_size,
                          double r_cut) {
    int N = static_cast<int>(pos.size());
    double half_dt2 = 0.5 * dt * dt;
    double half_dt  = 0.5 * dt;

    // positions update: r(t+dt) = r(t) + v(t)*dt + 0.5*a(t)*dt^2
    for (int i = 0; i < N; ++i) {
        for (int d = 0; d < 3; ++d) {
            pos[i][d] += vel[i][d] * dt + acc[i][d] * half_dt2;
        }
    }

    // save old accelerations, compute new
    std::vector<Vec3> acc_old = acc;
    compute_accelerations(pos, cell_size, r_cut, acc);

    // velocities update: v(t+dt) = v(t) + 0.5*(a(t) + a(t+dt))*dt
    for (int i = 0; i < N; ++i) {
        for (int d = 0; d < 3; ++d) {
            vel[i][d] += half_dt * (acc_old[i][d] + acc[i][d]);
        }
    }

    // wrap positions into [0, cell_size)
    for (int i = 0; i < N; ++i) {
        for (int d = 0; d < 3; ++d) {
            if (pos[i][d] >= cell_size) pos[i][d] -= cell_size;
            if (pos[i][d] < 0.0)       pos[i][d] += cell_size;
        }
    }
}

double kinetic_energy(const std::vector<Vec3>& vel) {
    double ek = 0.0;
    for (const auto& v : vel) {
        ek += v[0] * v[0] + v[1] * v[1] + v[2] * v[2];
    }
    return 0.5 * ek;
}

double potential_energy(const std::vector<Vec3>& pos,
                        double cell_size,
                        double r_cut) {
    int N = static_cast<int>(pos.size());
    double rc = std::min(r_cut, cell_size / 2.0);
    double rc2 = rc * rc;
    double half_cell = cell_size / 2.0;
    double u_shift = lj_potential(rc);
    double pe = 0.0;

    for (int i = 0; i < N; ++i) {
        for (int j = i + 1; j < N; ++j) {
            double dx = pos[j][0] - pos[i][0];
            double dy = pos[j][1] - pos[i][1];
            double dz = pos[j][2] - pos[i][2];

            if (dx >  half_cell) dx -= cell_size;
            if (dx < -half_cell) dx += cell_size;
            if (dy >  half_cell) dy -= cell_size;
            if (dy < -half_cell) dy += cell_size;
            if (dz >  half_cell) dz -= cell_size;
            if (dz < -half_cell) dz += cell_size;

            double r2 = dx * dx + dy * dy + dz * dz;
            if (r2 < rc2) {
                double r = std::sqrt(r2);
                pe += lj_potential(r) - u_shift;
            }
        }
    }
    return pe;
}
